### PR TITLE
[SYCL] Add support for [[intel::reqd_sub_group_…

### PR DESCRIPTION
**Pull request by smanna12**
…size()]]

This patch adds new spelling of IntelReqdSubGroupSize attribute
and generates a diagnostic which will tell that previous spelling
of the attribute is deprecated.

Signed-off-by: Soumi Manna <soumi.manna@intel.com>

### CODE REVIEW

File: clang/include/clang/Basic/DiagnosticSemaKinds.td [hunk 1]
```diff
@@ -10969,6 +10969,9 @@ def err_ivdep_declrefexpr_arg : Error<
 def warn_ivdep_redundant : Warning <"ignoring redundant Intel FPGA loop "
   "attribute 'ivdep': safelen %select{INF|%1}0 >= safelen %select{INF|%3}2">,
   InGroup<IgnoredAttributes>;
+def warn_attribute_spelling_deprecated : Warning<
+  "previous spelling of attribute %0 is deprecated">,
```
Comment (erichkeane) [inline clang/include/clang/Basic/DiagnosticSemaKinds.td:10973]: User's aren't particularly knowledgeable about the fact that the same attribute can be spelled in 2 different ways.  They tend to treat them as different attributes with similar/same behavior.

Additionally, we likely want to suggest 

This should probably be:
"attribute %0 is deprecated
Then a note of: <note_function_suggestion > "did you mean %1?"

Additionally, See if you can find if there is a good note that we can use to suggest

Comment (smanna12) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td:10973]: > User's aren't particularly knowledgeable about the fact that the same attribute can be spelled in 2 different ways. They tend to treat them as different attributes with similar/same behavior.
> 
> Additionally, we likely want to suggest
> 
> This should probably be:
> "attribute %0 is deprecated
> Then a note of: <note_function_suggestion > "did you mean %1?"
> 
> Additionally, See if you can find if there is a good note that we can use to suggest

Thanks @erichkeane  for the suggestion. I will look at this and update the diagnostic.

File: clang/include/clang/Basic/DiagnosticSemaKinds.td [hunk 2]
```diff
@@ -10969,6 +10969,9 @@ def err_ivdep_declrefexpr_arg : Error<
 def warn_ivdep_redundant : Warning <"ignoring redundant Intel FPGA loop "
   "attribute 'ivdep': safelen %select{INF|%1}0 >= safelen %select{INF|%3}2">,
   InGroup<IgnoredAttributes>;
+def warn_attribute_spelling_deprecated : Warning<
+  "attribute %0 is deprecated, did you mean to use 'reqd_sub_group_size' instead?">,
```
Comment (erichkeane) [inline clang/include/clang/Basic/DiagnosticSemaKinds.td:10973]: I had suggested making this a warning + note (using something like note_function_suggestion) as it seems to fit how diagnostics work better, and still have a preference for it.  But @elizabethandrews @Fznamznon and @premanandrao can make the final call here.

Comment (elizabethandrews) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td:10973]: I personally prefer adding a note. However I see instances of 'did you mean' in error/warning message itself in DiagnosticsSemaKinds.td. @erichkeane what does clang community recommend? If they recommend using a separate note, @smanna12 would you mind adding a note? There are examples in DiagnosticsSemaKinds.td

Comment (erichkeane) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td:10973]: I suspect we are consistently inconsistent :)   Clang typically doesn't have its warnings be prescriptive, but I can see that we do in a bunch of cases.  I'm not sure we have official guidance.

That said, if we leave it alone, i think the suggested replacement needs to be a replacement as well.  Finally, don't we mean:

'did you mean to use 'intel::reqd_sub_group_size' instead?"

File: clang/include/clang/Basic/DiagnosticSemaKinds.td [hunk 3]
```diff
@@ -10970,8 +10970,10 @@ def warn_ivdep_redundant : Warning <"ignoring redundant Intel FPGA loop "
   "attribute 'ivdep': safelen %select{INF|%1}0 >= safelen %select{INF|%3}2">,
   InGroup<IgnoredAttributes>;
 def warn_attribute_spelling_deprecated : Warning<
-  "attribute %0 is deprecated, did you mean to use 'reqd_sub_group_size' instead?">,
+  "attribute %0 is deprecated">,
   InGroup<DeprecatedAttributes>;
+def note_spelling_suggestion : Note<
+  "did you mean to use 'intel::reqd_sub_group_size' instead?">;
```
Comment (erichkeane) [inline clang/include/clang/Basic/DiagnosticSemaKinds.td:10976]: Make intel::reqd_sub_group_size a %0 instead, then fill it in with your Diag call.

File: clang/include/clang/Basic/DiagnosticSemaKinds.td (hunk 2 above)
Comment (smanna12) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td]: Thanks @elizabethandrews and @erichkeane for the comments. 

>> However I see instances of 'did you mean' in error/warning message itself in DiagnosticsSemaKinds.td.
Yes, i saw several examples that used and so added in the warning message. 

I have added warning+ note and updated all tests with it.

File: clang/include/clang/Basic/DiagnosticSemaKinds.td (hunk 3 above)
Comment (elizabethandrews) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td:10976]: I was actually surprised a general 'did you mean' like this one did not exist already.

Comment (erichkeane) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td:10976]: @elizabethandrews : the only one I found was note_function_suggestion, which is textually identical to what I'd like, but is function-specific.  I'm tempted to update the name in the upstream to make it work for anything...

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 4]
```diff
@@ -3014,6 +3014,13 @@ static void handleSubGroupSize(Sema &S, Decl *D, const ParsedAttr &AL) {
   if (D->getAttr<IntelReqdSubGroupSizeAttr>())
     S.Diag(AL.getLoc(), diag::warn_duplicate_attribute) << AL;
 
+  if (AL.getAttributeSpellingListIndex() ==
+      IntelReqdSubGroupSizeAttr::CXX11_cl_intel_reqd_sub_group_size) {
```
Comment (erichkeane) [inline clang/lib/Sema/SemaDeclAttr.cpp:3018]: Last question, I promise:)  DO we also intend to deprecate the GCC spelling (as intel_reqd_sub_group_size) ?  It seems silly (though not my call) to deprecate 1 but not the other.

File: clang/include/clang/Basic/DiagnosticSemaKinds.td (hunk 3 above)
Comment (smanna12) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td]: > Make intel::reqd_sub_group_size a %0 instead, then fill it in with your Diag call.

Done.

File: clang/test/SemaSYCL/sycl-device-reqd-sub-group-size-template.cpp [hunk 5]
```diff
@@ -5,7 +5,9 @@
 template <int SIZE>
 class KernelFunctor {
 public:
```
Comment (elizabethandrews) [inline clang/test/SemaSYCL/sycl-device-reqd-sub-group-size-template.cpp:7]: It feels weird to me that existing tests all throw warnings now. I am not sure what is usually done when spellings are changed though. Personally I would just prefer updating these tests to use the new spelling (except where this functionality is actually being checked). @erichkeane thoughts?

Comment (erichkeane) [reply clang/test/SemaSYCL/sycl-device-reqd-sub-group-size-template.cpp:7]: I could go either way.  There aren't that many tests to just 'fix' the spelling of here, so I'd probably suggest just fixing all the spellings instead.

If this was more widespread (that is, the tests were a pain to change), or a situation where there was a difference in the implementation, I'd consider just using the -Wno flag in the tests.  In this case, I think just changing the spelling in the tests is better than adding the warning/note to ALL of them.

Comment (erichkeane) [reply clang/test/SemaSYCL/sycl-device-reqd-sub-group-size-template.cpp:7]: That said, probably leave at least 1 to validate the deprecated warning :)

Comment (elizabethandrews) [reply clang/test/SemaSYCL/sycl-device-reqd-sub-group-size-template.cpp:7]: Yea. That's what I meant by 'except where this functionality is actually being checked' :) 

@smanna12 sorry but can you make this change? I think it's just cleaner this way.

Comment (smanna12) [reply clang/test/SemaSYCL/sycl-device-reqd-sub-group-size-template.cpp:7]: > Yea. That's what I meant by 'except where this functionality is actually being checked' :)
> 
> @smanna12 sorry but can you make this change? I think it's just cleaner this way.

Sure. I will do this.

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 4 above)
Comment (smanna12) [reply clang/lib/Sema/SemaDeclAttr.cpp:3018]: It is alight Erich. 

>>DO we also intend to deprecate the GCC spelling (as intel_reqd_sub_group_size) ?

My understanding is here that no, we do not intend to deprecate the GCC spelling (as intel_reqd_sub_group_size).

File: clang/test/SemaSYCL/sycl-device-reqd-sub-group-size-template.cpp (hunk 5 above)
Comment (erichkeane) [reply clang/test/SemaSYCL/sycl-device-reqd-sub-group-size-template.cpp:7]: I think you still need to leave ONE example (perhaps add a new example) to validate the new warning/note.

Comment (smanna12) [reply clang/test/SemaSYCL/sycl-device-reqd-sub-group-size-template.cpp:7]: yes, right. By mistake, I changed all spellings.

Comment (smanna12) [reply clang/test/SemaSYCL/sycl-device-reqd-sub-group-size-template.cpp:7]: Sorry for the multiple submission.

**Comment by smanna12** [comment]
Thanks for the review.

File: clang/test/SemaSYCL/reqd-sub-group-size-device.cpp [hunk 6]
```diff
@@ -1,22 +1,29 @@
...
 public:
+  // expected-warning@+2 {{attribute 'intel_reqd_sub_group_size' is deprecated}}
+  // expected-note@+1 {{did you mean to use 'intel::reqd_sub_group_size' instead?}}
   [[cl::intel_reqd_sub_group_size(16)]] void operator()() {}
 };
 
 class Functor8 { // expected-error {{conflicting attributes applied to a SYCL kernel}}
 public:
-  [[cl::intel_reqd_sub_group_size(8)]] void operator()() { // expected-note {{conflicting attribute is here}}
+  [[intel::reqd_sub_group_size(8)]] void operator()() { // expected-note {{conflicting attribute is here}}
```
Comment (MrSidims) [inline clang/test/SemaSYCL/reqd-sub-group-size-device.cpp:17]: I suppose support for a 'right side' spelling will go in another patch, correct?

File: clang/include/clang/Basic/AttrDocs.td [hunk 7]
```diff
@@ -3495,6 +3496,19 @@ as in the examples below:
        /* kernel code */
   });
 
+  class Functor
+  {
+      void operator()(item<1> item) [[intel::reqd_sub_group_size(16)]]
```
Comment (MrSidims) [inline clang/include/clang/Basic/AttrDocs.td:3501]: This example isn't covered by a test.

Comment (MrSidims) [reply clang/include/clang/Basic/AttrDocs.td:3501]: I mean that is exactly a 'right side' spelling. AFAIK clang isn't support it, considering an attribute to be a function type attribute, instead of function declaration, so I'm not expecting it to work 'out of the box'. Hence I'm not asking to add a test/or to add the appropriate code to support this spelling (as well as template instantiation) in a scope of this patch, but please correct the doc appropriately.

File: clang/test/SemaSYCL/reqd-sub-group-size-device.cpp (hunk 6 above)
Comment (smanna12) [reply clang/test/SemaSYCL/reqd-sub-group-size-device.cpp:17]: @MrSidims - Support for a "right side" attribute spelling is a separate issue, so this will go in another patch (PR).

 There are other attributes like SYCL_INTEL_reqd_work_group and  bunch of FPGA-related attributes i believe currently do not support "right side" spelling.   This is in on my task list and i will work on the support in another PR.

File: clang/include/clang/Basic/AttrDocs.td (hunk 7 above)
Comment (smanna12) [reply clang/include/clang/Basic/AttrDocs.td]: @MrSidims .  Thanks for the review and explanation. I have corrected Doc since this PR does not handle "right side" attribute spelling and we will add this support in a separate PR.

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 4 above)
Comment (smanna12) [reply clang/lib/Sema/SemaDeclAttr.cpp:3018]: This comment is showing unresolved. @erichkeane Do you have still have question on this?

Comment (erichkeane) [reply clang/lib/Sema/SemaDeclAttr.cpp:3018]: No, I don't.
